### PR TITLE
Rweber/zkp

### PR DESCRIPTION
**Pull request by rickwebiii**
Refactor API to facilitate ZKPs

### CODE REVIEW

File: sunscreen/src/fhe_compiler.rs [hunk 1]
```diff
@@ -46,117 +47,85 @@ pub trait FheProgramFn {
...
         Self::new()
     }
 }
 
-impl Compiler {
+/**
+ * A private method for converting between Compiler types. We don't want
+ * to expose this as part of the public API, hence we don't `impl From`.
+ */
+fn into<T, U>(x: Compiler<T>) -> Compiler<U> {
```
Comment (samtay) [inline sunscreen/src/fhe_compiler.rs:70]: Does it make sense to have this exist for all `T, U`? It's internal so it's not a huge deal, but if only certain conversions make sense, we could restrict this at the type level.

File: sunscreen_docs/src/advanced/batching/dot_product.md [hunk 2]
```diff
@@ -250,7 +250,7 @@ fn main() {
 
     // 3. Make our runtime, generate keys, encrypt our data, run our
     // program, and check the result
-    let runtime = Runtime::new(&fhe_program.metadata.params).unwrap();
+    let runtime = Runtime::new_fhe(&fhe_program.metadata.params).unwrap();
```
Comment (rickwebiii) [inline sunscreen_docs/src/advanced/batching/dot_product.md:253]: Got through each md file and make sure the reading is still consistent with the examples.

File: sunscreen/src/lib.rs [hunk 3]
```diff
@@ -105,39 +110,104 @@ impl Application {
...
+        Self {
+            fhe_programs: other.fhe_programs,
+            zkp_programs: other.zkp_programs,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl Application<FheZkp> {
+    pub(crate) fn to_application(other: Application<()>) -> Application<FheZkp> {
```
Comment (samtay) [inline sunscreen/src/lib.rs:205]: These impls seem odd. It seems like anytime someone uses `Application::new()` they get back an `app: Application<()>` and then need to use `Application<Something>::to_application(app)`? I imagine we can simplify this. 

In general I wonder if these `Struct<()>` impls with `new` and whatnot could instead just be generic. Then someone can specify what they want by just specifying the generic, e.g. `Application<Fhe>::new().get_fhe_program()`.

File: sunscreen_runtime/src/runtime.rs [hunk 4]
```diff
@@ -20,21 +22,77 @@ enum Context {
...
+    fn unwrap_fhe(&self) -> &FheRuntimeData {
+        match self {
+            Self::Fhe(x) => x,
+            _ => panic!("Expected RuntimeData::Fhe."),
+        }
+    }
+}
+
+/**
+ * Contains all the elements needed to encrypt, decrypt, generate keys, and evaluate FHE programs.
```
Comment (samtay) [inline sunscreen_runtime/src/runtime.rs:85]: This comment indicates the type is specific to FHE programs, but the type looks more general than that.

File: examples/amm/src/main.rs [hunk 5]
```diff
@@ -20,17 +21,17 @@ struct Miner {
...
-    runtime: Runtime,
+    runtime: FheRuntime,
 }
 
 impl Miner {
     pub fn setup() -> Result<Miner, Error> {
         let app = Compiler::new().fhe_program(swap_nu).compile()?;
 
-        let runtime = Runtime::new(app.params())?;
+        let runtime = GenericRuntime::new_fhe(app.params())?;
```
Comment (samtay) [inline examples/amm/src/main.rs:31]: All of these examples import `GenericRuntime`, but the docs in that module all reference the type synonym `Runtime`. I think the examples and docs should be in agreement on the intended use. (Personally I'd keep it as `Runtime` throughout.)

File: sunscreen/src/fhe_compiler.rs (hunk 1 above)
Comment (rickwebiii) [reply sunscreen/src/fhe_compiler.rs:70]: The cases this needs to handle are:
`Compiler<()> -> Compiler<Zkp>`
`Compiler<()> -> Compiler<Fhe>`
`Compiler<Zkp> -> Compiler<FheZkp>`
`Compiler<Fhe> -> Compiler<FheZkp>`

To add bounds to make all of these work, I think we would need to add a new trait and not use (), but some other marker type. I don't think it's worth it.

This function is indeed internal only.

File: sunscreen/src/lib.rs (hunk 3 above)
Comment (rickwebiii) [reply sunscreen/src/lib.rs:205]: This should be an `into()` function as we did with Compiler. But the idea is the same: our code internally needs to convert between types when appropriate, but users never do.

Users never construct Applications themselves (the `new` method is marked pub(crate)). Rather, the compiler emits the appropriately typed one.

**Comment by samtay** [comment]
Also, reminder to update the playground as part of this PR. Here's a [branch](https://github.com/Sunscreen-tech/rust-playground/compare/sunscreen...Sunscreen-tech:rust-playground:samtay/rw-zkp) with the current state of the `pir` example if you wanna just use that commit.

File: examples/amm/src/main.rs (hunk 5 above)
Comment (rickwebiii) [reply examples/amm/src/main.rs:31]: This was a bad F2 refactor. I've fixed this.

File: sunscreen_docs/src/advanced/batching/dot_product.md (hunk 2 above)
Comment (rickwebiii) [reply sunscreen_docs/src/advanced/batching/dot_product.md:253]: As discussed in our sync, we'll do this after checking in.

File: sunscreen/src/lib.rs (hunk 3 above)
Comment (rickwebiii) [reply sunscreen/src/lib.rs]: I've migrated to the same pattern as into with a method on

```
impl<T> Application<T> {
  fn convert<U>(self) -> Application<U> { ... }
}
```

**Pull request merged**